autoload/phpactor/project/projectRepository.vim

```diff
@@ -0,0 +1,47 @@
+function! phpactor#project#projectRepository#create() abort
+  return {
+        \ 'projects': {},
+        \ 'isEmpty': function('s:isEmpty'),
+        \ 'addProject': function('s:addProject'),
+        \ 'findProjectContainingFile': function('s:findProjectContainingFile'),
+        \ 'listProjects': function('s:listProjects'),
+        \ 'hasProjectWithPrimaryRoot': function('s:hasProjectWithPrimaryRoot'),
+        \ 'hasProject': function('s:hasProject')
+        \ }
+endfunction
+
+function! s:listProjects() dict abort
+  return copy(self.projects)
+endfunction
+
+function! s:findProjectContainingFile(file) dict abort
+    for l:rootPath in keys(self.projects)
+      let l:project = self.projects[l:rootPath]
+      if l:project.containsFile(a:file)
+        return l:project
+      endif
+    endfor
+
+    return v:null
+endfunction
+
+function! s:addProject(project) dict abort
+  " todo check type
+  if v:null isnot get(self.projects, a:project.getPrimaryRootPath(), v:null)
+    throw printf('Project "%s" already exists', a:project.getPrimaryRootPath())
+  endif
+
+  let self.projects[a:project.getPrimaryRootPath()] = a:project
```

Comment:
Given a project does not contain any data it's on no consequences if I try to register the same project twice, isn't it ?
If I'm not mistaken then I propose to fail silently to prevent user experience:
```suggestion
  " todo check type, here you could throw is the structure is not right, all keys must exist

  if get(self.projects, a:project.getPrimaryRootPath(), v:null) is v:null
    let self.projects[a:project.getPrimaryRootPath()] = a:project
  endif
```